packages/gen-ai/frontend/src/app/context/MCPContext.tsx

```diff
@@ -0,0 +1,188 @@
+import * as React from 'react';
+import { Namespace, useBrowserStorage } from 'mod-arch-core';
+import { MCPServerFromAPI, MCPToolsStatus, MCPServerConfig, TokenInfo } from '~/app/types';
+import { useMCPServers, ServerStatusInfo } from '~/app/hooks/useMCPServers';
+import { getMCPServerTools } from '~/app/services/llamaStackService';
+import { MCP_SERVERS_SESSION_STORAGE_KEY } from '~/app/utilities/const';
+
+type MCPContextValue = {
+  servers: MCPServerFromAPI[];
+  serversLoaded: boolean;
+  serversLoadError: Error | null;
+  serverStatuses: Map<string, ServerStatusInfo>;
+  statusesLoading: Set<string>;
+  allStatusesChecked: boolean;
+  playgroundSelectedServerIds: string[];
+  saveSelectedServersToPlayground: (serverIds: string[]) => void;
+  selectedServersCount: number;
+  setSelectedServersCount: (count: number) => void;
+  refresh: () => void;
+  fetchServerTools: (serverUrl: string, mcpBearerToken?: string) => Promise<MCPToolsStatus>;
+  serverTokens: Map<string, TokenInfo>;
+  setServerTokens: React.Dispatch<React.SetStateAction<Map<string, TokenInfo>>>;
+  isServerValidated: (serverUrl: string) => boolean;
+  getSelectedServersForAPI: () => MCPServerConfig[];
+  checkServerStatus: (serverUrl: string, mcpBearerToken?: string) => Promise<ServerStatusInfo>;
+};
+
+const MCPContext = React.createContext<MCPContextValue | null>(null);
+
+type MCPContextProviderProps = {
+  children: React.ReactNode;
+  namespace: Namespace | undefined;
+};
+
+export const MCPContextProvider: React.FunctionComponent<MCPContextProviderProps> = ({
+  children,
+  namespace,
+}) => {
+  const selectedProject = namespace?.name;
+
+  const {
+    servers,
+    serversLoaded,
+    serversLoadError,
+    serverStatuses,
+    statusesLoading,
+    allStatusesChecked,
+    refresh,
+    checkServerStatus,
+  } = useMCPServers(selectedProject || '', { autoCheckStatuses: false });
+
+  const [playgroundSelectedServerIds, setPlaygroundSelectedServerIds] = useBrowserStorage<string[]>(
+    MCP_SERVERS_SESSION_STORAGE_KEY,
+    [],
+    true,
+    true,
+  );
+
+  const [selectedServersCount, setSelectedServersCount] = React.useState<number>(0);
+  const [serverTokens, setServerTokens] = React.useState<Map<string, TokenInfo>>(new Map());
+
+  React.useEffect(() => {
+    setServerTokens(new Map());
+  }, [selectedProject]);
+
+  const saveSelectedServersToPlayground = React.useCallback(
+    (serverIds: string[]) => {
+      setPlaygroundSelectedServerIds(serverIds);
+    },
+    [setPlaygroundSelectedServerIds],
+  );
+
+  const fetchServerTools = React.useCallback(
```

Comment:
Why is there a util function inside a context? This could easily be shared as a normal function.